File: docs/Sword2offer/CH4-解题思路/Python/checkStack.py
class Stack:
  def __init__(self):
    self.items = []

  def append(self, value):
    self.items.append(value)

  def pop(self):
    return self.items.pop()

  def isEmpty(self):
    return len(self.items) == 0

  def top(self):
    return self.items[-1]


def check(inPath, outPath):
  stack = Stack()
  in_index = 0
  out_index = 0
  while in_index < len(inPath):
    stack.append(inPath[in_index])
    if stack.top() == outPath[out_index]:
      stack.pop()
      out_index += 1
    in_index += 1
  while not stack.isEmpty():
    if stack.top() != outPath[out_index]:
      return False
    else:
      stack.pop()
      out_index += 1
  return True

if __name__ == '__main__':
  in_arr = [1, 2, 3, 4, 5]
  out_arr = [4, 5, 3, 2, 1]
  out_arr2 = [4, 3, 5, 1, 2]
  print(check(in_arr, out_arr))
  print(check(in_arr, out_arr2))

File: docs/Sword2offer/CH4-解题思路/ES5/checkStack.js
"use strict";

function Stack() {
  var items = [];
  this.push = function (value) {
    items.push(value);
  };
  this.pop = function () {
    return items.pop();
  };
  this.top = function () {
    return items[items.length - 1];
  };
  this.isEmpty = function () {
    return items.length === 0;
  };
  this.size = function () {
    return items.length;
  };
  this.clear = function () {
    items = [];
  };
  this.show = function () {
    console.log(items);
  };
}

var check = function check(inPath, outPath) {
  var stack = new Stack();
  // 压入inpath元素
  while (inPath.length) {
    stack.push(inPath.shift());
    // 如果和弹出对应,就pop
    if (stack.top() === outPath[0]) {
      stack.pop();
      outPath.shift();
    }
  }
  // 如果stack不为空, 继续和outpath检测
  while (!stack.isEmpty()) {
    // 只要有一个不相等,就要是false
    if (stack.top() !== outPath[0]) {
      return false;
    }
    stack.pop();
    outPath.shift();
  }
  return true;
};

var inArr = [1, 2, 3, 4, 5];
var outArr = [4, 5, 3, 2, 1];
// 由于直接使用了shift会修改原数组,所以重新创建了一份
var inArr2 = [1, 2, 3, 4, 5];
var outArr2 = [4, 3, 5, 1, 2];
console.log(check(inArr, outArr));
console.log(check(inArr2, outArr2));